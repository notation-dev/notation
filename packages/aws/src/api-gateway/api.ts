import { AwsResourceGroup } from "@notation/aws.iac/client";
import * as aws from "@notation/aws.iac/resources";

export const api = (rgConfig: { name: string }) => {
  const apiGroup = new AwsResourceGroup("api", rgConfig);

  const apiResource = apiGroup.add(
    new aws.apiGateway.Api({
      config: {
        Name: rgConfig.name,
        ProtocolType: "HTTP",
      },
    }),
  );

  apiGroup.add(
    new aws.apiGateway.Stage({
      config: { StageName: "$default", AutoDeploy: true },
      dependencies: { api: apiResource },
    }),
  );

  return apiGroup;
};
