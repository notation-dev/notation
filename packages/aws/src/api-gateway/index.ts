export * from "./api";
export * from "./route";
export * from "./router";
export * from "./auth";
