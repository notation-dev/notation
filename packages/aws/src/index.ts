export const apiRoute = (...args: any) => {};
export const handler = (...args: any) => {};

export type FnConfig = any;
