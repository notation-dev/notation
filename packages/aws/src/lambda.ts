export type FnConfig = {
  service: "aws/lambda";
  memory?: number;
  timeout?: number;
};
