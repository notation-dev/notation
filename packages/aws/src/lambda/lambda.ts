import * as aws from "@notation/aws.iac";
import * as std from "@notation/std.iac";
import crypto from "crypto";
import path from "path";

type LambdaConfig = {
  id?: string;
  handler: string;
  code: {
    type: "file" | "zip";
    path: string;
  };
  runtime?: aws.lambda.LambdaFunctionConfig["Runtime"];
};

export const lambda = (config: LambdaConfig): aws.AwsResourceGroup => {
  const functionGroup = new aws.AwsResourceGroup("Lambda", { config });
  const filePath = config.code.path;

  let lambdaId = config.id;

  if (!lambdaId) {
    const filePathHash = crypto
      .createHash("BLAKE2s256")
      .update(filePath)
      .digest("hex")
      .slice(0, 8);

    lambdaId = `${config.handler}-${filePathHash}`;
  }

  let zipFile: std.fs.ZipFileInstance | std.fs.FileInstance;

  if (config.code.type === "file") {
    zipFile = functionGroup.add(
      new std.fs.Zip({
        id: `${lambdaId}-zip`,
        config: { sourceFilePath: filePath },
      }),
    );
  } else {
    zipFile = functionGroup.add(
      new std.fs.File({
        id: `${lambdaId}-zip`,
        config: { filePath },
      }),
    );
  }

  const role = functionGroup.add(
    new aws.lambda.LambdaIamRole({
      id: `${lambdaId}-role`,
      config: {
        RoleName: `${lambdaId}-role`,
      },
    }),
  );

  functionGroup.add(
    new aws.lambda.LambdaRolePolicyAttachment({
      id: `${lambdaId}-policy`,
      dependencies: { role },
    }),
  );

  const fileName = path.parse(filePath).name;
  const runtime = config.runtime || "nodejs22.x";

  const lambdaResource = functionGroup.add(
    new aws.lambda.LambdaFunction({
      id: lambdaId,
      config: {
        FunctionName: lambdaId,
        Handler: `${fileName}.${config.handler}`,
        Runtime: runtime,
        // todo: make this configurable and remove it as a default
        ReservedConcurrentExecutions: 1,
      },
      dependencies: {
        role,
        zipFile,
      },
    }),
  );

  functionGroup.add(
    new aws.lambda.LambdaLogGroup({
      id: `${lambdaId}-log-group`,
      config: { retentionInDays: 30 },
      dependencies: { lambda: lambdaResource },
    }),
  );

  return functionGroup;
};
