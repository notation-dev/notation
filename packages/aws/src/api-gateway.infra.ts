import { AwsResourceGroup } from "@notation/aws.iac/client";
import { apiGateway, lambda } from "@notation/aws.iac/resources";
import { ApiGatewayHandler, fn } from "./lambda";

export const api = (rgConfig: { name: string }) => {
  const apiGroup = new AwsResourceGroup("api", rgConfig);

  const apiResource = apiGroup.add(
    new apiGateway.Api({
      config: {
        Name: rgConfig.name,
        ProtocolType: "HTTP",
      },
    }),
  );

  apiGroup.add(
    new apiGateway.Stage({
      config: { StageName: "$default", AutoDeploy: true },
      dependencies: { router: apiResource },
    }),
  );

  return apiGroup;
};

export const route = (
  apiGroup: ReturnType<typeof api>,
  method: string, // todo: http methods only
  path: `/${string}`,
  handler: ApiGatewayHandler,
) => {
  const apiResource = apiGroup.findResource(
    "aws/api-gateway",
  ) as apiGateway.ApiInstance;

  // at compile time becomes infra module
  const fnGroup = handler as any as ReturnType<typeof fn>;

  const routeGroup = new AwsResourceGroup("api/route", {
    dependencies: { router: apiGroup.id, fn: fnGroup.id },
  });

  let integration;

  const lambdaResource = fnGroup.findResource(
    "aws/lambda",
  )! as lambda.LambdaInstance;

  const permission = fnGroup.findResource("aws/lambda/permission/api-gateway");

  integration = fnGroup.findResource(
    "aws/api-gateway/integration/lambda",
  ) as apiGateway.LambdaIntegrationInstance;

  if (!permission) {
    fnGroup.add(
      new lambda.LambdaApiGatewayPermission({
        dependencies: {
          api: apiResource,
          lambda: lambdaResource,
        },
      }),
    );
  }

  if (!integration) {
    integration = fnGroup.add(
      new apiGateway.LambdaIntegration({
        dependencies: {
          api: apiResource,
          lambda: lambdaResource,
        },
      }),
    );
  }

  routeGroup.add(
    new apiGateway.Route({
      config: {
        RouteKey: `${method} ${path}`,
      },
      dependencies: {
        api: apiResource,
        lambdaIntegration: integration,
      },
    }),
  );

  return routeGroup;
};
