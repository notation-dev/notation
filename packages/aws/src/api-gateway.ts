import { handle } from "./lambda";

export const apiRoute = (...args: any) => {};

export const handler = handle.apiGateway;

export const json = (result: any) => ({
  body: JSON.stringify(result),
  statusCode: 200,
});
