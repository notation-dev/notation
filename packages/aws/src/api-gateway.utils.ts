import { ApiGatewayHandler } from "./lambda";
import { api, route } from "./api-gateway.infra";

export const router = (apiGroup: ReturnType<typeof api>) => {
  const createRouteCallback =
    (method: string) => (path: `/${string}`, handler: ApiGatewayHandler) => {
      return route(apiGroup, method, path, handler);
    };

  return {
    get: createRouteCallback("GET"),
    post: createRouteCallback("POST"),
    put: createRouteCallback("PUT"),
    patch: createRouteCallback("PATCH"),
    delete: createRouteCallback("DELETE"),
  };
};

export const json = (result: any) => ({
  body: JSON.stringify(result),
  statusCode: 200,
});
