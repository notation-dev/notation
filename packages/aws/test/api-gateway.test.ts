import { test, expect } from "bun:test";
import { json } from "src/api-gateway";
import { api, route, router, json } from "src/api-gateway";
import { fn } from "src/lambda";

test("json returns a JSON string and a 200 status code", () => {
  const payload = { message: "Hello, world!" };
  const response = json(payload);
  expect(response.statusCode).toEqual(200);
  expect(response.body).toEqual(JSON.stringify(payload));
});

test("api resource group snapshot", () => {
  const apiResourceGroup = api({ name: "api" });
  expect(apiResourceGroup).toMatchSnapshot();
});

test("route resource group snapshot", () => {
  const apiResourceGroup = api({ name: "api" });
  const fnResourceGroup = fn({ handler: "handler.fn.js" });

  const routeResourceGroup = route(
    apiResourceGroup,
    "GET",
    "/hello",
    fnResourceGroup as any,
  );

  expect(routeResourceGroup).toMatchSnapshot();
  expect(fnResourceGroup).toMatchSnapshot();
});

test("route resource group idempotency snapshot", () => {
  const apiResourceGroup = api({ name: "api" });
  const fnResourceGroup = fn({ handler: "handler.fn.js" });

  route(apiResourceGroup, "GET", "/hello", fnResourceGroup as any);
  const fnResourceGroupSnapshot = JSON.stringify(fnResourceGroup);
  route(apiResourceGroup, "POST", "/hello", fnResourceGroup as any);
  const fnResourceGroupSnapshot2 = JSON.stringify(fnResourceGroup);

  expect(fnResourceGroupSnapshot).toEqual(fnResourceGroupSnapshot2);
});

