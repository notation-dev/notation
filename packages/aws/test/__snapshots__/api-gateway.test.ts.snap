// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`api resource group snapshot 1`] = `
AwsResourceGroup {
  "config": {
    "name": "api",
  },
  "dependencies": {},
  "id": 0,
  "platform": "aws",
  "resources": [
    {
      "config": {
        "Name": "api",
        "ProtocolType": "HTTP",
      },
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {},
      "getPrimaryKey": [Function],
      "groupId": 0,
      "id": 0,
      "output": null,
      "read": [Function: AsyncFunction],
      "retryOn": undefined,
      "type": "aws/api-gateway",
      "update": [Function: AsyncFunction],
    },
    {
      "config": {
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {
        "api": {
          "config": {
            "Name": "api",
            "ProtocolType": "HTTP",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {},
          "getPrimaryKey": [Function],
          "groupId": 0,
          "id": 0,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": undefined,
          "type": "aws/api-gateway",
          "update": [Function: AsyncFunction],
        },
      },
      "getPrimaryKey": [Function],
      "groupId": 0,
      "id": 1,
      "output": null,
      "read": [Function: AsyncFunction],
      "retryOn": undefined,
      "type": "aws/api-gateway/stage",
      "update": [Function: AsyncFunction],
    },
  ],
  "type": "api",
}
`;

exports[`route resource group snapshot 1`] = `
AwsResourceGroup {
  "config": {},
  "dependencies": {
    "fn": 1,
    "router": 0,
  },
  "id": 2,
  "platform": "aws",
  "resources": [
    {
      "config": {
        "RouteKey": "GET /hello",
      },
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {
        "api": {
          "config": {
            "Name": "api",
            "ProtocolType": "HTTP",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {},
          "getPrimaryKey": [Function],
          "groupId": 0,
          "id": 0,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": undefined,
          "type": "aws/api-gateway",
          "update": [Function: AsyncFunction],
        },
        "lambdaIntegration": {
          "config": {},
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {
            "api": {
              "config": {
                "Name": "api",
                "ProtocolType": "HTTP",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {},
              "getPrimaryKey": [Function],
              "groupId": 0,
              "id": 0,
              "output": null,
              "read": [Function: AsyncFunction],
              "retryOn": undefined,
              "type": "aws/api-gateway",
              "update": [Function: AsyncFunction],
            },
            "lambda": {
              "config": {
                "FunctionName": "function-1",
                "Handler": "index.handler.fn.js",
                "Runtime": "nodejs18.x",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {
                "policyAttachment": {
                  "config": {
                    "PolicyArn": "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
                  },
                  "create": [Function: AsyncFunction],
                  "delete": [Function: AsyncFunction],
                  "dependencies": {
                    "role": {
                      "config": {
                        "RoleName": "1-role",
                      },
                      "create": [Function: AsyncFunction],
                      "delete": [Function: AsyncFunction],
                      "dependencies": {},
                      "getPrimaryKey": [Function],
                      "groupId": 1,
                      "id": 3,
                      "output": null,
                      "read": [Function: AsyncFunction],
                      "retryOn": undefined,
                      "type": "aws/lambda/role",
                      "update": [Function: AsyncFunction],
                    },
                  },
                  "getPrimaryKey": [Function],
                  "groupId": 1,
                  "id": 4,
                  "output": null,
                  "read": undefined,
                  "retryOn": undefined,
                  "type": "aws/lambda/policy-attachment",
                  "update": undefined,
                },
                "role": {
                  "config": {
                    "RoleName": "1-role",
                  },
                  "create": [Function: AsyncFunction],
                  "delete": [Function: AsyncFunction],
                  "dependencies": {},
                  "getPrimaryKey": [Function],
                  "groupId": 1,
                  "id": 3,
                  "output": null,
                  "read": [Function: AsyncFunction],
                  "retryOn": undefined,
                  "type": "aws/lambda/role",
                  "update": [Function: AsyncFunction],
                },
                "zipFile": {
                  "config": {
                    "filePath": "src/fns/handler.fn.js",
                  },
                  "create": [Function: AsyncFunction],
                  "delete": [Function: AsyncFunction],
                  "dependencies": {},
                  "getPrimaryKey": [Function],
                  "groupId": 1,
                  "id": 2,
                  "output": null,
                  "read": [Function: AsyncFunction],
                  "retryOn": undefined,
                  "type": "std/zip",
                  "update": [Function: AsyncFunction],
                },
              },
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 5,
              "output": null,
              "read": [Function: AsyncFunction],
              "retryOn": [
                "InvalidParameterValueException: The role defined for the function cannot be assumed by Lambda",
              ],
              "type": "aws/lambda",
              "update": [Function: AsyncFunction],
            },
          },
          "getPrimaryKey": [Function],
          "groupId": 1,
          "id": 8,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": undefined,
          "type": "aws/api-gateway/integration/lambda",
          "update": [Function: AsyncFunction],
        },
      },
      "getPrimaryKey": [Function],
      "groupId": 2,
      "id": 9,
      "output": null,
      "read": [Function: AsyncFunction],
      "retryOn": undefined,
      "type": "aws/api-gateway/route",
      "update": [Function: AsyncFunction],
    },
  ],
  "type": "api/route",
}
`;

exports[`route resource group snapshot 2`] = `
AwsResourceGroup {
  "config": {
    "config": {
      "fileName": "src/fns/handler.fn.js",
      "handler": "handler.fn.js",
    },
  },
  "dependencies": {},
  "id": 1,
  "platform": "aws",
  "resources": [
    {
      "config": {
        "filePath": "src/fns/handler.fn.js",
      },
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {},
      "getPrimaryKey": [Function],
      "groupId": 1,
      "id": 2,
      "output": null,
      "read": [Function: AsyncFunction],
      "retryOn": undefined,
      "type": "std/zip",
      "update": [Function: AsyncFunction],
    },
    {
      "config": {
        "RoleName": "1-role",
      },
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {},
      "getPrimaryKey": [Function],
      "groupId": 1,
      "id": 3,
      "output": null,
      "read": [Function: AsyncFunction],
      "retryOn": undefined,
      "type": "aws/lambda/role",
      "update": [Function: AsyncFunction],
    },
    {
      "config": {
        "PolicyArn": "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
      },
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {
        "role": {
          "config": {
            "RoleName": "1-role",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {},
          "getPrimaryKey": [Function],
          "groupId": 1,
          "id": 3,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": undefined,
          "type": "aws/lambda/role",
          "update": [Function: AsyncFunction],
        },
      },
      "getPrimaryKey": [Function],
      "groupId": 1,
      "id": 4,
      "output": null,
      "read": undefined,
      "retryOn": undefined,
      "type": "aws/lambda/policy-attachment",
      "update": undefined,
    },
    {
      "config": {
        "FunctionName": "function-1",
        "Handler": "index.handler.fn.js",
        "Runtime": "nodejs18.x",
      },
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {
        "policyAttachment": {
          "config": {
            "PolicyArn": "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {
            "role": {
              "config": {
                "RoleName": "1-role",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {},
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 3,
              "output": null,
              "read": [Function: AsyncFunction],
              "retryOn": undefined,
              "type": "aws/lambda/role",
              "update": [Function: AsyncFunction],
            },
          },
          "getPrimaryKey": [Function],
          "groupId": 1,
          "id": 4,
          "output": null,
          "read": undefined,
          "retryOn": undefined,
          "type": "aws/lambda/policy-attachment",
          "update": undefined,
        },
        "role": {
          "config": {
            "RoleName": "1-role",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {},
          "getPrimaryKey": [Function],
          "groupId": 1,
          "id": 3,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": undefined,
          "type": "aws/lambda/role",
          "update": [Function: AsyncFunction],
        },
        "zipFile": {
          "config": {
            "filePath": "src/fns/handler.fn.js",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {},
          "getPrimaryKey": [Function],
          "groupId": 1,
          "id": 2,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": undefined,
          "type": "std/zip",
          "update": [Function: AsyncFunction],
        },
      },
      "getPrimaryKey": [Function],
      "groupId": 1,
      "id": 5,
      "output": null,
      "read": [Function: AsyncFunction],
      "retryOn": [
        "InvalidParameterValueException: The role defined for the function cannot be assumed by Lambda",
      ],
      "type": "aws/lambda",
      "update": [Function: AsyncFunction],
    },
    {
      "config": {},
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {
        "lambda": {
          "config": {
            "FunctionName": "function-1",
            "Handler": "index.handler.fn.js",
            "Runtime": "nodejs18.x",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {
            "policyAttachment": {
              "config": {
                "PolicyArn": "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {
                "role": {
                  "config": {
                    "RoleName": "1-role",
                  },
                  "create": [Function: AsyncFunction],
                  "delete": [Function: AsyncFunction],
                  "dependencies": {},
                  "getPrimaryKey": [Function],
                  "groupId": 1,
                  "id": 3,
                  "output": null,
                  "read": [Function: AsyncFunction],
                  "retryOn": undefined,
                  "type": "aws/lambda/role",
                  "update": [Function: AsyncFunction],
                },
              },
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 4,
              "output": null,
              "read": undefined,
              "retryOn": undefined,
              "type": "aws/lambda/policy-attachment",
              "update": undefined,
            },
            "role": {
              "config": {
                "RoleName": "1-role",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {},
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 3,
              "output": null,
              "read": [Function: AsyncFunction],
              "retryOn": undefined,
              "type": "aws/lambda/role",
              "update": [Function: AsyncFunction],
            },
            "zipFile": {
              "config": {
                "filePath": "src/fns/handler.fn.js",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {},
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 2,
              "output": null,
              "read": [Function: AsyncFunction],
              "retryOn": undefined,
              "type": "std/zip",
              "update": [Function: AsyncFunction],
            },
          },
          "getPrimaryKey": [Function],
          "groupId": 1,
          "id": 5,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": [
            "InvalidParameterValueException: The role defined for the function cannot be assumed by Lambda",
          ],
          "type": "aws/lambda",
          "update": [Function: AsyncFunction],
        },
      },
      "getPrimaryKey": [Function],
      "groupId": 1,
      "id": 6,
      "output": null,
      "read": undefined,
      "retryOn": undefined,
      "type": "aws/lambda/log-group",
      "update": undefined,
    },
    {
      "config": {},
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {
        "api": {
          "config": {
            "Name": "api",
            "ProtocolType": "HTTP",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {},
          "getPrimaryKey": [Function],
          "groupId": 0,
          "id": 0,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": undefined,
          "type": "aws/api-gateway",
          "update": [Function: AsyncFunction],
        },
        "lambda": {
          "config": {
            "FunctionName": "function-1",
            "Handler": "index.handler.fn.js",
            "Runtime": "nodejs18.x",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {
            "policyAttachment": {
              "config": {
                "PolicyArn": "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {
                "role": {
                  "config": {
                    "RoleName": "1-role",
                  },
                  "create": [Function: AsyncFunction],
                  "delete": [Function: AsyncFunction],
                  "dependencies": {},
                  "getPrimaryKey": [Function],
                  "groupId": 1,
                  "id": 3,
                  "output": null,
                  "read": [Function: AsyncFunction],
                  "retryOn": undefined,
                  "type": "aws/lambda/role",
                  "update": [Function: AsyncFunction],
                },
              },
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 4,
              "output": null,
              "read": undefined,
              "retryOn": undefined,
              "type": "aws/lambda/policy-attachment",
              "update": undefined,
            },
            "role": {
              "config": {
                "RoleName": "1-role",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {},
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 3,
              "output": null,
              "read": [Function: AsyncFunction],
              "retryOn": undefined,
              "type": "aws/lambda/role",
              "update": [Function: AsyncFunction],
            },
            "zipFile": {
              "config": {
                "filePath": "src/fns/handler.fn.js",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {},
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 2,
              "output": null,
              "read": [Function: AsyncFunction],
              "retryOn": undefined,
              "type": "std/zip",
              "update": [Function: AsyncFunction],
            },
          },
          "getPrimaryKey": [Function],
          "groupId": 1,
          "id": 5,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": [
            "InvalidParameterValueException: The role defined for the function cannot be assumed by Lambda",
          ],
          "type": "aws/lambda",
          "update": [Function: AsyncFunction],
        },
      },
      "getPrimaryKey": [Function],
      "groupId": 1,
      "id": 7,
      "output": null,
      "read": undefined,
      "retryOn": undefined,
      "type": "aws/lambda/permission/api-gateway",
      "update": undefined,
    },
    {
      "config": {},
      "create": [Function: AsyncFunction],
      "delete": [Function: AsyncFunction],
      "dependencies": {
        "api": {
          "config": {
            "Name": "api",
            "ProtocolType": "HTTP",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {},
          "getPrimaryKey": [Function],
          "groupId": 0,
          "id": 0,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": undefined,
          "type": "aws/api-gateway",
          "update": [Function: AsyncFunction],
        },
        "lambda": {
          "config": {
            "FunctionName": "function-1",
            "Handler": "index.handler.fn.js",
            "Runtime": "nodejs18.x",
          },
          "create": [Function: AsyncFunction],
          "delete": [Function: AsyncFunction],
          "dependencies": {
            "policyAttachment": {
              "config": {
                "PolicyArn": "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {
                "role": {
                  "config": {
                    "RoleName": "1-role",
                  },
                  "create": [Function: AsyncFunction],
                  "delete": [Function: AsyncFunction],
                  "dependencies": {},
                  "getPrimaryKey": [Function],
                  "groupId": 1,
                  "id": 3,
                  "output": null,
                  "read": [Function: AsyncFunction],
                  "retryOn": undefined,
                  "type": "aws/lambda/role",
                  "update": [Function: AsyncFunction],
                },
              },
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 4,
              "output": null,
              "read": undefined,
              "retryOn": undefined,
              "type": "aws/lambda/policy-attachment",
              "update": undefined,
            },
            "role": {
              "config": {
                "RoleName": "1-role",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {},
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 3,
              "output": null,
              "read": [Function: AsyncFunction],
              "retryOn": undefined,
              "type": "aws/lambda/role",
              "update": [Function: AsyncFunction],
            },
            "zipFile": {
              "config": {
                "filePath": "src/fns/handler.fn.js",
              },
              "create": [Function: AsyncFunction],
              "delete": [Function: AsyncFunction],
              "dependencies": {},
              "getPrimaryKey": [Function],
              "groupId": 1,
              "id": 2,
              "output": null,
              "read": [Function: AsyncFunction],
              "retryOn": undefined,
              "type": "std/zip",
              "update": [Function: AsyncFunction],
            },
          },
          "getPrimaryKey": [Function],
          "groupId": 1,
          "id": 5,
          "output": null,
          "read": [Function: AsyncFunction],
          "retryOn": [
            "InvalidParameterValueException: The role defined for the function cannot be assumed by Lambda",
          ],
          "type": "aws/lambda",
          "update": [Function: AsyncFunction],
        },
      },
      "getPrimaryKey": [Function],
      "groupId": 1,
      "id": 8,
      "output": null,
      "read": [Function: AsyncFunction],
      "retryOn": undefined,
      "type": "aws/api-gateway/integration/lambda",
      "update": [Function: AsyncFunction],
    },
  ],
  "type": "aws/function",
}
`;
